privacyidea/lib/tokens/tiqrtoken.py

```diff
@@ -440,6 +440,8 @@ def check_challenge_response(self, user=None, passw=None, options=None):
             for challengeobject in challengeobject_list:
                 # check if we are still in time.
                 if challengeobject.is_valid():
+                    # This is a valid transaction ID not need to increase fail counter
+                    self.set_failcount(abs(self.get_failcount()) - 1)
```

Comment:
Thanks a lot for your pull request!

It somehow feels a bit wired to decrease the failcounter. I am not sure, which side effects this might have.
I would rather not increase the failcounter in the first place, if possible.

The failcounter is increased in 3 occasions:

If the pin matches, but the otp or challenge/response does not match:
https://github.com/privacyidea/privacyidea/blob/master/privacyidea/lib/token.py#L2266

If the pin does not match, aka nothing matches:
https://github.com/privacyidea/privacyidea/blob/master/privacyidea/lib/token.py#L2293

Or as in your case: the response does not match a challenge:
https://github.com/privacyidea/privacyidea/blob/master/privacyidea/lib/token.py#L2240

If the tokenclass would know, in which case the inc_failcount occurs, than the tokenclass itself could decide in which way to increase the failcounter.
We could add a parameter to inc_failcount to pass the situation or we could add methods to the tokenclass.

What do you think?
@Mipronimo What do you think in regards to our push token? I totally forgot about the tiqr, which works rather similar but checks the answered challenge in another way.

Reply:
I agree that it is a bit awkward to decrease the failcounter.  Can the `message` then also be adjusted?:
 * `"message": "Response did not match the challenge."`

Maybe to: `message: "User has not yet confirmed"`

Reply:
I would like to redesign this a bit as mentioned above to also bring this in sync with the push token, since TiQR and push work rather the same.
Currently we are a bit under heavy load, so please excuse, if this takes some time.

Reply:
I mentioned it now in #1689, which should be discuss here maybe..
We could add the parameter `transaction_id` to the /auth endpoint.
After `/auth username=user password=pass` we get a transaction id.
If we then poll `/auth username=user transaction_id=tID` we would not increase the failcounter for the other tokens, right?

Reply:
"polling" `/auth` with will currently always increase the fail counter.

The interesting thing however is, that the second /auth or /validate/check request as an answer to the first request in a challenge-response-flow would not actually result in a successful authentication. So "Polling" the /auth oder /validate/check endpoint will not enable brute force for tiqr or push, since the authentication occurs out of band.

Basically it happens here:
https://github.com/privacyidea/privacyidea/blob/master/privacyidea/lib/token.py#L2240
This code increases the fail counters of the tokens, that match the challenge, if none of the tokens created a valid response. This is fine.
But not for tiqr and push. So it should be something like this:

A)
~~~~python
        if not res:
            # We did not find any successful response, so we need to increase the
            # failcounters
            for token_obj in challenge_response_token_list:
                if token_obj.get_type().lower() not in ["tiqr", "push"]:
                    token_obj.inc_failcount()
~~~~

But this of course is bad style, since we should avoid having tokenspecific stuff in the lib/token.py.
So either we add a flag to the tokenclass like ``is_out_of_band_response_token``:
B)
~~~~python
            for token_obj in challenge_response_token_list:
                if not token_obj.is_out_of_band_response_token:
                    token_obj.inc_failcount()
~~~~

or we add a new method, that is increases the token and is only called here.
C)
~~~~python
            for token_obj in challenge_response_token_list:
                    token_obj.inc_failcount_in_case_of_response()
~~~~
But I think it is a bit strange to have a method, that is only called in one place and either increases or does nothing.

So I think the token-flag **B)** might be a way to go...